filebeat/crawler/registrar.go

```diff
@@ -212,28 +199,15 @@ func (r *Registrar) getPreviousFile(newFilePath string, newFileInfo os.FileInfo)
 }
 
 func (r *Registrar) setState(path string, state *FileState) {
-	r.stateMutex.Lock()
-	defer r.stateMutex.Unlock()
 	r.state[path] = state
 }
 
 func (r *Registrar) getState(path string) (*FileState, bool) {
-	r.stateMutex.Lock()
-	defer r.stateMutex.Unlock()
 	state, exist := r.state[path]
 	return state, exist
 }
 
-func (r *Registrar) updateStateSource(state *FileState, path string) {
-	r.stateMutex.Lock()
-	defer r.stateMutex.Unlock()
-	state.Source = path
-}
-
 func (r *Registrar) getStateCopy() map[string]FileState {
-	r.stateMutex.Lock()
-	defer r.stateMutex.Unlock()
-
 	copy := make(map[string]FileState)
 	for k, v := range r.state {
 		copy[k] = *v
```

Comment:
I'd say both. Looks like you want to rely on copy-on-read to prevent race conditions, so you need the mutex to prevent writes during the copy operation and you need a deep copy of `FileState`. Changing `FileState` to contain a `FileStateOS` instead of `*FileStateOS` will make it simple to create a deep copy.